gadget/raw.go

```diff
@@ -0,0 +1,281 @@
+ *
+ */
+package gadget
+
+import (
+	"bytes"
+	"crypto"
+	_ "crypto/sha1"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// RawStructureWriter implements support for writing raw (bare) structures.
+type RawStructureWriter struct {
+	rootDir string
+}
+
+// NewRawStructureWriter returns a structure writer that will load the structure
+// and content data from provided root directory.
+func NewRawStructureWriter(rootDir string) *RawStructureWriter {
+	return &RawStructureWriter{rootDir: rootDir}
+}
+
+// WriteRawStream writes the input stream in that corresponds to provided
+// positioned content. The number of bytes read from input stream must match
+// exactly the declared size of positioned content entry.
+func (r *RawStructureWriter) WriteRawStream(out io.WriteSeeker, pc *PositionedContent, in io.Reader) error {
+	if _, err := out.Seek(int64(pc.StartOffset), io.SeekStart); err != nil {
+		return fmt.Errorf("cannot seek to content start offset 0x%x: %v", pc.StartOffset, err)
+	}
+
+	_, err := io.CopyN(out, in, int64(pc.Size))
+	if err != nil {
+		return fmt.Errorf("cannot write image: %v", err)
+	}
+	return nil
+}
+
+// WriteRawImage writes a single image described by a positioned content entry.
+func (r *RawStructureWriter) WriteRawImage(out io.WriteSeeker, pc *PositionedContent) error {
+	img, err := os.Open(filepath.Join(r.rootDir, pc.Image))
+	if err != nil {
+		return fmt.Errorf("cannot open image file: %v", err)
+	}
+	defer img.Close()
+
+	return r.WriteRawStream(out, pc, img)
+}
+
+// Write will write the whole contents of a bare structure ps into the output stream.
+func (r *RawStructureWriter) Write(out io.WriteSeeker, ps *PositionedStructure) error {
+	if !ps.IsBare() {
+		return fmt.Errorf("structure %v is not bare", ps)
+	}
+	for _, pc := range ps.PositionedContent {
+		if err := r.WriteRawImage(out, &pc); err != nil {
+			return fmt.Errorf("failed to write image %v: %v", pc, err)
+		}
+	}
+	return nil
+}
+
+// RawStructureUpdater implements support for updating raw (bare) structures.
+type RawStructureUpdater struct {
+	rootDir      string
+	backupDir    string
+	deviceLookup locationLookupFunc
+}
+
+type locationLookupFunc func(ps *PositionedStructure) (string, error)
+
+// NewRawStructureUpdater returns an updater for raw (bare) structures. Update
+// data will be loaded from provided root directory. Backups of replaced
+// structures are temporarily kept in the rollback directory.
+func NewRawStructureUpdater(rootDir, backupDir string, deviceLookup locationLookupFunc) *RawStructureUpdater {
+	return &RawStructureUpdater{
+		rootDir:      rootDir,
+		backupDir:    backupDir,
+		deviceLookup: deviceLookup,
+	}
+}
+
+func (r *RawStructureUpdater) contentBackupPath(ps *PositionedStructure, pc *PositionedContent, idx int) string {
+	return filepath.Join(r.backupDir, fmt.Sprintf("struct-%v-%v-%v", ps.Index, idx, encodeLabel(pc.String())))
```

Comment:
to me it feels like pc etc should have methods that do encodeLabel on the relevant materials, instead of being the responsability of the callers.

Reply:
It kind of does, there's the String() method. OTOH, `encodeLabel()` is just a heleper that works like blkid_encode_string. Maybe I should just drop the string part and keep around the index given content had in gadget.yaml, just like it was done for the structure (aka. PositionedStructure).

Reply:
Pushed some changes in #6912.